planner/core/rule_join_reorder_greedy.go

```diff
@@ -0,0 +1,203 @@
+		copy(eqConds, p.EqualConditions)
+		copy(otherConds, p.OtherConditions)
+		return []LogicalPlan{lChild, rChild}, eqConds, otherConds
+	}
+	lhsJoinGroup, eqEdges, otherConds := extractInnerJoinGroup(lJoin)
+	if lhsJoinGroup == nil {
+		return nil, nil, nil
+	}
+	return append(lhsJoinGroup, rChild), append(eqEdges, p.EqualConditions...), append(otherConds, p.OtherConditions...)
+}
+
+type joinReOrderGreedySolver struct {
+	ctx          sessionctx.Context
+	curJoinGroup []LogicalPlan
+	eqEdges      []*expression.ScalarFunction
+	otherConds   []expression.Expression
+}
+
+func (s *joinReOrderGreedySolver) solve() (LogicalPlan, error) {
+	for _, node := range s.curJoinGroup {
+		_, err := node.deriveStats()
+		if err != nil {
+			return nil, err
+		}
+	}
+	sort.Slice(s.curJoinGroup, func(i, j int) bool {
+		return s.curJoinGroup[i].statsInfo().RowCount < s.curJoinGroup[j].statsInfo().RowCount
+	})
+
+	var cartesianGroup []LogicalPlan
+	for len(s.curJoinGroup) > 0 {
+		newNode, err := s.enlargeJoinTree()
+		if err != nil {
+			return nil, err
+		}
+		cartesianGroup = append(cartesianGroup, newNode)
+	}
+
+	return s.makeBushyJoin(cartesianGroup), nil
+}
+
+func (s *joinReOrderGreedySolver) enlargeJoinTree() (LogicalPlan, error) {
+	curJoinTree := s.curJoinGroup[0]
+	s.curJoinGroup = s.curJoinGroup[1:]
+	for {
+		bestCost := math.MaxFloat64
+		bestIdx := -1
+		var finalRemainOthers []expression.Expression
+		var bestJoin LogicalPlan
+		for i, node := range s.curJoinGroup {
+			newJoin, remainOthers := s.connectedAndNewNode(curJoinTree, node)
+			if newJoin == nil {
+				continue
+			}
+			_, err := newJoin.deriveStats()
+			if err != nil {
+				return nil, err
+			}
+			curCost := curJoinTree.statsInfo().RowCount + newJoin.statsInfo().RowCount + node.statsInfo().RowCount
+			if bestCost > curCost {
+				bestCost = curCost
+				bestJoin = newJoin
+				bestIdx = i
+				finalRemainOthers = remainOthers
+			}
+		}
+		if bestJoin == nil {
+			break
+		}
+		curJoinTree = bestJoin
+		s.curJoinGroup = append(s.curJoinGroup[:bestIdx], s.curJoinGroup[bestIdx+1:]...)
+		s.otherConds = finalRemainOthers
+	}
+	return curJoinTree, nil
+}
+
+func (s *joinReOrderGreedySolver) connectedAndNewNode(leftNode, rightNode LogicalPlan) (LogicalPlan, []expression.Expression) {
+	var usedEdges []*expression.ScalarFunction
+	remainOtherConds := make([]expression.Expression, len(s.otherConds))
+	copy(remainOtherConds, s.otherConds)
+	for _, edge := range s.eqEdges {
+		lCol := edge.GetArgs()[0].(*expression.Column)
+		rCol := edge.GetArgs()[1].(*expression.Column)
+		if leftNode.Schema().Contains(lCol) && rightNode.Schema().Contains(rCol) {
+			usedEdges = append(usedEdges, edge)
+		} else if rightNode.Schema().Contains(lCol) && leftNode.Schema().Contains(rCol) {
+			newSf := expression.NewFunctionInternal(s.ctx, ast.EQ, edge.GetType(), rCol, lCol).(*expression.ScalarFunction)
+			usedEdges = append(usedEdges, newSf)
```

Comment:
why not directly use `edge`?

Reply:
We need to make sure that the eq conds is in form of `col from left eq col from right`.

Reply:
Uh.. Why should we guarantee this property? Is there any later operation relies on this property?

Reply:
Yes. Otherwise, if we want to use the left join key of the equal condition, we need to get the arg and check it in left side's schema.